presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -195,4 +245,76 @@ private void initializeColumnReaders()
             columnReadersMap.put(column, ParquetColumnReader.createReader(column));
         }
     }
+
+    private Block readBlock(String name, Type type, List<String> path, List<Integer> offsets, List<Integer> positions)
+            throws IOException
+    {
+        path.add(name);
+        Optional<RichColumnDescriptor> descriptor = getDescriptor(fileSchema, requestedSchema, path);
+        if (!descriptor.isPresent()) {
+            path.remove(name);
+            if (positions.isEmpty()) {
+                return RunLengthEncodedBlock.create(type, null, batchSize);
+            }
+            return RunLengthEncodedBlock.create(type, null, positions.size());
+        }
+
+        Block block = null;
+        if (ROW.equals(type.getTypeSignature().getBase())) {
+            block = readStruct(type, path, positions);
+            getBlockOffsets(block, offsets);
+        }
+        else if (MAP.equals(type.getTypeSignature().getBase())) {
+            block = readMap(type, path, positions);
+            getBlockOffsets(block, offsets);
+        }
+        else if (ARRAY.equals(type.getTypeSignature().getBase())) {
+            block = readArray(type, path, positions);
+            getBlockOffsets(block, offsets);
+        }
+        else {
+            block = readPrimitive(descriptor.get(), type, offsets, positions);
+        }
+        path.remove(name);
+        return block;
+    }
+
+    private void getBlockOffsets(Block block, List<Integer> offsets)
+    {
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            offsets.add(block.getLength(i));
+        }
+    }
+
+    private List<Integer> getNonNullPositions(int positionCount, Block keys, List<Integer> offsets)
+    {
+        List<Integer> positions = new ArrayList(keys.getPositionCount());
```

Comment:
missing diamond ` List<Integer> positions = new ArrayList<>(keys.getPositionCount());` (again IntelliJ should warn you)